File: SWEA/src/Q17675_민석이의_마니또/solution.cpp
#include <iostream>
#include <queue>
#include <algorithm>
#include <string.h>
#define f first
#define s second
using namespace std;
int N, M, visited[404], ans;
vector<vector<pair<int, int>>> v;

void input() {
    cin >> N >> M;
    v = vector<vector<pair<int, int>>>(N + 1);
    int n1, n2, d;
    for (int i = 0; i < M; i++) {
        cin >> n1 >> n2 >> d;
        v[n1].push_back({d, n2});
    }
}

void bfs(int start) {
    priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> pq;
    pq.push({0, start});

    while (!pq.empty()) {
        pair<int, int> curr = pq.top();
        pq.pop();
        for (auto n: v[curr.s]) {
            if (visited[n.s]) continue;
            visited[n.s]++;
            if (start == n.s) ans = min(ans, (curr.f + n.f));
            pq.push({(curr.f + n.f), n.s});
        }
    }
}

void solve() {
    ans = 987654321;
    for (int i = 1; i <= N; i++) {
        memset(visited, 0, sizeof(visited));
        bfs(i);
    }
}

void output(int t) {
    cout << "#" << t << " " << (ans == 987654321 ? -1 : ans) << "\n";
}

int main() {
    ios::sync_with_stdio(0);
    cin.tie(0), cout.tie(0);
    int T;
    cin >> T;
    for (int t = 1; t <= T; t++) {
        input();
        solve();
        output(t);
    }
}

File: SWEA/src/Q17675_민석이의_마니또/Solution.java
package Q17675_민석이의_마니또;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.*;

public class Solution {
    static BufferedReader br;
    static StringTokenizer stk;
    static StringBuilder sb;
    static int N, M, ans = 0, visited[];


    static class Node {
        int next;
        int dist;

        Node(int next, int dist) {
            this.next = next;
            this.dist = dist;
        }
    }
    static PriorityQueue<Node> pq;
    static List<Node> list[];

    private static void input() throws Exception {
        stk = new StringTokenizer(br.readLine());
        N = Integer.parseInt(stk.nextToken());
        M = Integer.parseInt(stk.nextToken());
        list = new ArrayList[N + 1];
        ans = 987654321;
        for (int i = 1; i <= N; i++) {
            list[i] = new ArrayList<>();
        }
        int n1, n2, d;
        for (int i = 0; i < M; i++) {
            stk = new StringTokenizer(br.readLine());
            n1 = Integer.parseInt(stk.nextToken());
            n2 = Integer.parseInt(stk.nextToken());
            d = Integer.parseInt(stk.nextToken());
            list[n1].add(new Node(n2, d));
        }
    }

    private static void bfs(int start) {
        pq = new PriorityQueue<>((n1, n2) -> n1.dist - n2.dist);
        pq.add(new Node(start, 0));
        while (!pq.isEmpty()) {
            Node curr = pq.poll();
            for (Node n : list[curr.next]) {
                if (visited[n.next] > 0) continue;
                visited[n.next]++;
                if (start == n.next) ans = Math.min(ans, curr.dist + n.dist);
                pq.add(new Node(n.next, curr.dist + n.dist));
            }
        }
    }

    private static void solve() {
        for (int i = 1; i <= N; i++) {
            visited = new int[N + 1];
            bfs(i);
        }
    }

    private static void output(int t) {
        System.out.println();
    }

    public static void main(String[] args) throws Exception {
        br = new BufferedReader(new InputStreamReader(System.in));
        sb = new StringBuilder();
        int T = Integer.parseInt(br.readLine());
        for (int t = 1; t <= T; t++) {
            input();
            solve();
            output(t);
        }
        System.out.print(sb.toString());
    }
}
